src/Raven.Server/Smuggler/Documents/MultiShardedDestination.cs

```diff
@@ -0,0 +1,289 @@
+using System.Threading.Tasks;
+using Raven.Client.Documents.Operations.Counters;
+using Raven.Client.Documents.Smuggler;
+using Raven.Client.Util;
+using Raven.Server.Documents;
+using Raven.Server.Documents.ShardedHandlers;
+using Raven.Server.Documents.ShardedHandlers.ShardedCommands;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Smuggler.Documents.Data;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Server;
+using Sparrow.Threading;
+
+namespace Raven.Server.Smuggler.Documents
+{
+    public class MultiShardedDestination : ISmugglerDestination
+    {
+        private readonly ShardedContext _shardedContext;
+        private readonly ShardedSmugglerHandler _handler;
+        private readonly ISmugglerSource _source;
+        private readonly StreamDestination[] _destinations;
+        private DatabaseSmugglerOptionsServerSide _options;
+
+        public MultiShardedDestination(ISmugglerSource source, ShardedContext shardedContext, ShardedSmugglerHandler handler)
+        {
+            _source = source;
+            _shardedContext = shardedContext;
+            _handler = handler;
+            _destinations = new StreamDestination[shardedContext.ShardCount];
+        }
+
+        public async ValueTask<IAsyncDisposable> InitializeAsync(DatabaseSmugglerOptionsServerSide options, SmugglerResult result, long buildVersion)
+        {
+            _options = options;
+            var holders = new StreamDestinationHolder[_shardedContext.ShardCount];
+
+            var importOperation = new ShardedImportOperation(_handler, holders, options);
+            var t = _shardedContext.ShardExecutor.ExecuteParallelForAllAsync(importOperation);
+
+            await Task.WhenAll(importOperation.ExposedStreamTasks);
+
+            for (int i = 0; i < holders.Length; i++)
+            {
+                await PrepareShardStreamDestination(holders, i, result, buildVersion);
+            }
+
+            return new AsyncDisposableAction(async () =>
+            {
+                for (int i = 0; i < holders.Length; i++)
+                {
+                    await holders[i].DisposeAsync();
+                }
+
+                await t;
+            });
+        }
+
+        private async Task PrepareShardStreamDestination(StreamDestinationHolder[] holders, int shard, SmugglerResult result, long buildVersion)
+        {
+            var stream = _handler.GetOutputStream(holders[shard].OutStream.OutputStream.Result, _options);
+            holders[shard].InputStream = stream;
+            holders[shard].ContextReturn = _handler.ContextPool.AllocateOperationContext(out JsonOperationContext context);
+            var destination = new StreamDestination(stream, context, _source);
+            holders[shard].DestinationAsyncDisposable = await destination.InitializeAsync(_options, result, buildVersion);
+            _destinations[shard] = destination;
+        }
+
+        internal struct StreamDestinationHolder : IAsyncDisposable
```

Comment:
Why is that a struct? If this is `IAsyncDisposable`, it is going to be boxed for sure.